operational_analysis/toolkits/reanalysis_downloading.py

```diff
@@ -0,0 +1,316 @@
+        en_date_ds (:obj:`pandas.Timestamp`): The last datetime in the dataset
+
+    Returns:
+        (:obj:`pandas.Timestamp`, :obj:`pandas.Timestamp`): The final start and end datetimes
+    """
+
+    # Convert start and end dates to Timestamps if in another format
+    if st_date is not None:
+        st_date_new = pd.to_datetime(st_date)
+
+    # Add one hour since the last time step will not be retrieved
+    if en_date is not None:
+        en_date_new = pd.to_datetime(en_date) + datetime.timedelta(hours=1)
+
+    # Determine start and end dates if both are not defined
+    if (st_date is not None) & (en_date is None):
+        # Set end date to num_years years after start date
+
+        # Check to see if start date is out of bounds for the dataset
+        if st_date_new < st_date_ds:
+            print("Start date is out of range. Changing to " + str(st_date_ds))
+            st_date_new = st_date_ds
+
+        # Handle rare leap year case where start date is on the last day of February
+        try:
+            en_date_new = st_date_new.replace(year=st_date_new.year + num_years)
+        except ValueError:
+            en_date_new = st_date_new.replace(month=2, day=28, year=st_date_new.years + num_years)
+
+    elif (st_date is None) & (en_date is not None):
+        # Set start date to num_years years before end date
+
+        # Check to see if end date is out of bounds for the dataset
+        if (en_date_new - datetime.timedelta(hours=1)) > en_date_ds:
+            print("End date is out of range. Changing to " + str(en_date_ds))
+            en_date_new = en_date_ds + datetime.timedelta(hours=1)
+
+        # Handle rare leap year case where start date is on the last day of February
+        try:
+            st_date_new = en_date_new.replace(year=en_date_new.year - num_years)
+        except ValueError:
+            st_date_new = en_date_new.replace(month=2, day=28, year=en_date_new.years - num_years)
+
+    elif (st_date is None) & (en_date is None):
+        # Find end of last complete month and set start date to num_years before
+        # Check if last date is the end of a month
+
+        if en_date_ds.month == (en_date_ds + datetime.timedelta(hours=1)).month:
+            en_date_new = en_date_ds.replace(day=1, hour=0, minute=0)
+        else:
+            en_date_new = (en_date_ds + datetime.timedelta(hours=1)).replace(minute=0)
+
+        st_date_new = en_date_new.replace(year=en_date_new.year - num_years)
+    # else: start and end dates are defined already, so don't need to do anything
+
+    # Check once more to see if start and end dates are out of bounds for the dataset
+    if st_date_new < st_date_ds:
+        print("Start date is out of range. Changing to " + str(st_date_ds))
+        st_date_new = st_date_ds
+
+        # If start date minute changes, update end date too
+        en_date_new = en_date_new.replace(minute=st_date_new.minute)
```

Comment:
Adding num_years wouldn't necessarily work here because the user might have defined a specific end date they want to use. This would be a rare case, but the reason for this line is to make sure the difference between the start and end dates is an integer number of hours. Say the user specified start and end dates that are on the hour (e.g., 12:00). If the start date is out of bounds, and gets changed to the first date in the data set, which could be on a half hour (e.g., 00:30), then changing the end date's minute value to also be 30 minutes will make sure the PlanetOS request asks for the right number of total hours.